### PR TITLE
Add rewrite to merge consecutive `Join`ed `Subtensor`s

### PR DESCRIPTION
**Pull request by ricardoV94**
I was playing a bit with this kind of rewrites recently, which appeared for me when undoing `at.diff` operations which reduce the first axis index.

`concatenate(([x[:1], x[1:])) -> x`

I am not sure this is a useful rewrite to have in the codebase, let me know what you think.
If this is useful, it will probably still need a bit of cleanup, as I've never worked with SubTensors before.

### CODE REVIEW

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1673]: If you need to check that two term graphs are (syntactically) the same, use the function in `aesara.graph.basic`.  That will cover more cases, as well.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1673]: Thanks. I assume you mean `equal_computations`

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1689]:
```suggestion
    r"""Simplify contiguous :class:`Subtensor`\s inside a :class:`Join`.
```

Keep in mind that we need to be more consistent about the format and formatting of our docstrings.

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1725]: If you `continue` on the negation of this condition, you can simplify these and the following statements.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1685]: Yeah I am okay with not having this in canonicalization. As for the useless, I still didn't figure out when is that pass used or if it's simply an optional thing

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1687]:
```suggestion
    r"""Simplify contiguous :class:`Subtensor`\s inside a :class:`Join`.
```

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1689]:
```suggestion
        join((x[:3], x[3:5]), axis=0) -> x[:5]
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1678]: If the sub-graphs that evaluate to `True` under `equal_computations` can be merged by `MergeOptimizer`, then use of `equal_computations` can be wasteful (remember, it may need to walk the entire graph).  In such cases, it might be better to use identity comparisons and wait until the sub-graphs have been merged.

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1689]: pre-commit disagrees with that extra indentation

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1678]: That seems to work just fine! Should we still leave the rewrite as `specialize` only?

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [reply aesara/tensor/subtensor_opt.py:1678]: The indent in the docstring?  Weird.

Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1678]: pre-commit just overrides it. Let me see which hook is it

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [reply aesara/tensor/subtensor_opt.py]: > That seems to work just fine! Should we still leave the rewrite as `specialize` only?

It's not a question of whether `equal_computations` works, but one of whether or not it also works _without_ `equal_computations` and its associated costs.

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [reply aesara/tensor/subtensor_opt.py:1678]: Well, perhaps we can just add a `..code:` above it and get some formatting, or wrap it in backticks.

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py]: All the tests I had written worked out of the box...

Comment (brandonwillard) [reply aesara/tensor/subtensor_opt.py]: Without it?

Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py]: My question is whether this makes the rewrite light enough to justify being a canonicalization. I assumed that was the concern before for making it only a specialization

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1678]: Something like this?
```python
r"""Simplify contiguous :class:`Subtensor`\s inside a :class:`Join`.

..code:
  join((x[:3], x[3:5]), axis=0) -> x[:5]
"""
```

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py]: > Without it?

Yes. I am no longer using any `equal_computations` in the new rewrite

Comment (brandonwillard) [reply aesara/tensor/subtensor_opt.py]: This is a tough one, because it's a justifiable canonicalization (given our current standard, at least), and it's not completely clear which approach (i.e. `equal_computations`/no `equal_computations) is better on average&mdash;at least in terms of cost.

It's always possible that using `equal_computations` could avoid a few intermediate steps (e.g. the merge steps needed to make the identity-based comparisons `True`, and possibly more), but that has to be offset by the cost of using `equal_computations` in all other cases.

Comment (brandonwillard) [reply aesara/tensor/subtensor_opt.py]: As a default, I think it's better to work within the rewrite pipeline, and `equal_computations` has the potential to "override" the basic rewrite pass logic.  For example, given a fixed set of deterministically ordered rewrite passes, another rewrite might be expected to act upon a graph that would've been rewritten under the use of `equal_computations`.  In other words, use of `equal_computations` has the potential to preempt the set of rewrite passes that would've otherwise been applied, because it effectively performs a merge out-of-sequence.

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [reply aesara/tensor/subtensor_opt.py:1678]: I think it's `.. code::` (or `.. code:: python`).

**Review by brandonwillard — DISMISSED** [review]
A conflict needs to be resolved; otherwise, this looks good to merge.

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1749]: Why is `__getitem__` being called explicitly here?

File: aesara/tensor/subtensor_opt.py
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1736]: Why `is`?

Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1736]: IIRC, because we wanted to avoid use of `equal_computations` and rely on variable identity / constant merging

Comment (brandonwillard) [reply aesara/tensor/subtensor_opt.py:1736]: `==` is the best approach, because it requires fewer redundant comparisons across types that implement `__eq__` properly.

Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1736]: Oh, I was under the impression that equality of TensorVariables was always `True` (non-empty)!

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1749]: No good reason

**Comment by ricardoV94** [comment]
Fixed the wrong assumption that we can perform this rewrite as long as the steps are equal, got rid of `is` comparisons and simplified the tests.

File: aesara/tensor/subtensor_opt.py
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py]: :facepalm:

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**
